src/main/python/community-detector/community_detector.py

```diff
@@ -0,0 +1,154 @@
+    linear = (edges == "linear")
+    graphs = []
+    communities = []
+
+    if not linear:
+        log.debug("Transposing the matrix")
+        buckmat_csc = buckets_matrix.T.tocsr()
+
+    fat_ccs = []
+
+    for vertices in ccs.values():
+        if len(vertices) == 1:
+            continue
+        if len(vertices) == 2:
+            communities.append(vertices)
+            continue
+        fat_ccs.append(vertices)
+
+    log.debug("Building %d graphs", len(fat_ccs))
+
+    for vertices in fat_ccs:
+        if linear:
+            edges = []
+            weights = []
+            bucket_weights = buckets_matrix.sum(axis=0)
+            buckets = set()
+            for i in vertices:
+                for j in range(buckindptr[i], buckindptr[i + 1]):
+                    bucket = buckindices[j]
+                    weights.append(bucket_weights[0, bucket])
+                    bucket += total_nvertices
+                    buckets.add(bucket)
+                    edges.append((str(i), str(bucket)))
+        else:
+            edges = set()
+            weights = None
+            buckets = set()
+            for i in vertices:
+                for j in range(buckindptr[i], buckindptr[i + 1]):
+                    buckets.add(buckindices[j])
+            for bucket in buckets:
+                buckverts = \
+                    buckmat_csc.indices[buckmat_csc.indptr[bucket]:buckmat_csc.indptr[bucket + 1]]
+                for i, x in enumerate(buckverts):
+                    for y in buckverts:
+                        if x < y:
+                            edges.add((str(x), str(y)))
+            buckets.clear()
+            edges = list(edges)
+
+        graph = Graph(directed=False)
+        graph.add_vertices(list(map(str, vertices + list(buckets))))
+        graph.add_edges(edges)
+        graph.edge_weights = weights
+        graphs.append(graph)
+
+    log.debug("Launching the community detection")
+    detector = CommunityDetector(algorithm=algorithm, config=algorithm_params)
+
+    communities.extend(chain.from_iterable((detector(g) for g in graphs)))
+
+    log.debug("Overall communities: %d", len(communities))
+    log.debug("Average community size: %.1f",
+              numpy.mean([len(c) for c in communities]))
+    log.debug("Median community size: %.1f",
+              numpy.median([len(c) for c in communities]))
+    log.debug("Max community size: %d", max(map(len, communities)))
+
+    # Replaces CommunitiesModel().construct(communities, ccsmodel.id_to_element).save(output)
+    size = sum(map(len, communities))
+    data = numpy.zeros(size, dtype=numpy.uint32)
+    indptr = numpy.zeros(len(communities) + 1, dtype=numpy.int64)
+    pos = 0
+    for i, community in enumerate(communities):
+        data[pos:pos + len(community)] = community
+        pos += len(community)
+        indptr[i + 1] = pos
+
+    return {"data": data, "indptr": indptr}
+
+
+class CommunityDetector:
+    def __init__(self, algorithm, config):
```

Comment:
As discussed, it would be nice to have a references to origins of the code, coming from other projects i.e https://github.com/src-d/apollo/blob/0a6f445a799231f749a7d70610fb53e06ab02763/apollo/graph.py#L267 here.